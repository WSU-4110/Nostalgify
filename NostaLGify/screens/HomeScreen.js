import { StyleSheet, Text, View, TextInput, FlatList, TouchableOpacity, Image } from 'react-native';
import React, { useState, useEffect } from 'react';
import { AntDesign } from '@expo/vector-icons';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { FontAwesome6 } from '@expo/vector-icons';
import AsyncStorage from '@react-native-async-storage/async-storage';
import AccessTokenSingleton from './AccessTokenSingleton';

const accessTokenInstance = AccessTokenSingleton.getInstance();
accessTokenInstance.setAccessToken('your_access_token_here');

const accessToken = accessTokenInstance.getAccessToken();

async function fetchWebApi(endpoint, method, body, token) {
    const headers = {
        Authorization: `Bearer ${token}`,
      };
      let url = `https://api.spotify.com/${endpoint}`;
      if (method === 'GET' && body) {
        // If it's a GET request and there's a body, encode it into the URL
        const queryParams = new URLSearchParams(body).toString();
        url += `?${queryParams}`;
      }
      const res = await fetch(url, {
        headers,
        method,
        body: method !== 'GET' ? JSON.stringify(body) : undefined, // Only include body for non-GET requests
      });
      return await res.json();
  }
  
  async function getCurrentTrack(token) {
    // Endpoint reference: https://developer.spotify.com/documentation/web-api/reference/player/get-the-users-currently-playing-track/
    return await fetchWebApi(
      'v1/me/player/currently-playing', 'GET', null, token
    );
  }
  
  const HomeScreen = () => {
    const [currentTrack, setCurrentTrack] = useState(null);
  
    useEffect(() => {
      fetchCurrentTrack();
    }, []);
  
    const fetchCurrentTrack = async () => {
      try {
        const accessToken = await AsyncStorage.getItem('accessToken');
        if (accessToken) {
          const currentTrackData = await getCurrentTrack(accessToken);
          if (currentTrackData?.item) {
            setCurrentTrack(currentTrackData.item);
          } else {
            setCurrentTrack(null); // Set currentTrack to null if no track is playing
          }
        } else {
          console.error('Access token not found in AsyncStorage');
        }
      } catch (error) {
        console.error('Error fetching current track:', error);
      }
    };
  
    return (
      <View style={styles.container}>
        {currentTrack && (
          <View>
            <Image
              style={styles.albumCover}
              source={{ uri: currentTrack.album.images[0].url }}
            />
            <Text style={styles.trackName}>{currentTrack.name}</Text>
            <Text style={styles.artistName}>{currentTrack.artists.map(artist => artist.name).join(', ')}</Text>
          </View>
        )}
        {!currentTrack && (
          <Text style={styles.noTrack}>No track currently playing</Text>
        )}
      </View>
    );
  };
  
  const styles = StyleSheet.create({
    container: {
      flex: 1,
      backgroundColor: '#cca2b7',
      paddingTop: 50,
      alignItems: 'center',
      justifyContent: 'center',
    },
    albumCover: {
      width: 200,
      height: 200,
      marginBottom: 20,
    },
    trackName: {
      fontSize: 20,
      fontWeight: 'bold',
      marginBottom: 10,
    },
    artistName: {
      fontSize: 16,
    },
    noTrack: {
      fontSize: 18,
      color: 'white',
    },
  });
  
  export default HomeScreen;
