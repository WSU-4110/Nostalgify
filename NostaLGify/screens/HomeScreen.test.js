import React, { useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { fetchWebApi, getCurrentTrack, fetchCurrentTrack } from './HomeScreen';
import { render } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import HomeScreen from './HomeScreen'; // Adjust the path if needed

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
}));
jest.mock('react', () => ({
  ...jest.requireActual('react'), // Keep all actual exports from 'react'
  useState: jest.fn(), // Mock useState
}));
jest.mock('./HomeScreen', () => ({
  fetchWebApi: jest.fn(),
  getCurrentTrack: jest.fn(),
  fetchCurrentTrack: jest.fn(),
}));

beforeEach(() => {
  AsyncStorage.getItem.mockClear();
  fetchWebApi.mockClear();
  getCurrentTrack.mockClear();
  fetchCurrentTrack.mockClear();
});

describe('fetchWebApi', () => {
    test('should fetch data from Spotify API', async () => {
      // Mocking the fetch function
      global.fetch = jest.fn(() =>
        Promise.resolve({
          json: () => Promise.resolve({ data: 'mocked data' }),
        })
      );
  
      // Call the function
      const result = await fetchWebApi('endpoint', 'GET', null, 'token');
  
      // Assert
      expect(result).toEqual({ data: 'mocked data' });
      expect(fetch).toHaveBeenCalledWith(expect.any(String), {
        headers: expect.any(Object),
        method: 'GET',
      });
    });
  });

  describe('getCurrentTrack', () => {
    test('should fetch current track from Spotify API', async () => {
      // Mocking the fetchWebApi function
      const mockFetchWebApi = jest.fn(() => Promise.resolve({ item: { name: 'Mocked Track' } }));
      jest.spyOn(global, 'fetchWebApi').mockImplementation(mockFetchWebApi);
  
      // Call the function
      const result = await getCurrentTrack('token');
  
      // Assert
      expect(result).toEqual({ item: { name: 'Mocked Track' } });
      expect(mockFetchWebApi).toHaveBeenCalledWith('v1/me/player/currently-playing', 'GET', null, 'token');
    });
  });
  
  describe('fetchCurrentTrack', () => {
    test('should fetch current track and set state', async () => {
      // Mocking AsyncStorage.getItem and getCurrentTrack functions
      AsyncStorage.getItem = jest.fn(() => Promise.resolve('mockedToken'));
      const mockGetCurrentTrack = jest.fn(() => Promise.resolve({ item: { name: 'Mocked Track' } }));
      jest.spyOn(global, 'getCurrentTrack').mockImplementation(mockGetCurrentTrack);
  
      // Call the function
      await fetchCurrentTrack();
  
      // Assert
      expect(mockGetCurrentTrack).toHaveBeenCalledWith('mockedToken');
      expect(setCurrentTrack).toHaveBeenCalledWith({ name: 'Mocked Track' });
    });
  });

  describe('HomeScreen Component', () => {
    test('should render current track information if track is available', () => {
      const mockCurrentTrack = { album: { images: [{ url: 'mockedUrl' }] }, name: 'Mocked Track', artists: [{ name: 'Artist' }] };
      useState.mockReturnValueOnce([mockCurrentTrack, jest.fn()]);
  
      const { getByText, getByTestId } = render(<HomeScreen />);
  
      expect(getByTestId('albumCover').props.source.uri).toBe('mockedUrl');
      expect(getByText('Mocked Track')).toBeInTheDocument();
      expect(getByText('Artist')).toBeInTheDocument();
    });
  
    test('should render no track message if no track is available', () => {
      useState.mockReturnValueOnce([null, jest.fn()]);
  
      const { getByText } = render(<HomeScreen />);
  
      expect(getByText('No track currently playing')).toBeInTheDocument();
    });
  });

  describe('HomeScreen Styles', () => {
    test('should match snapshot', () => {
      const tree = renderer.create(<HomeScreen />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('fetchCurrentTrack Error Handling', () => {
    test('should log error if access token not found', async () => {
      AsyncStorage.getItem = jest.fn(() => Promise.resolve(null));
      console.error = jest.fn();
  
      await fetchCurrentTrack();
  
      expect(console.error).toHaveBeenCalledWith('Access token not found in AsyncStorage');
    });
  
    test('should log error if error occurs during track fetch', async () => {
      AsyncStorage.getItem = jest.fn(() => Promise.resolve('mockedToken'));
      global.console.error = jest.fn();
      getCurrentTrack.mockRejectedValueOnce('mockedError');
  
      await fetchCurrentTrack();
  
      expect(console.error).toHaveBeenCalledWith('Error fetching current track:', 'mockedError');
    });
  });
  