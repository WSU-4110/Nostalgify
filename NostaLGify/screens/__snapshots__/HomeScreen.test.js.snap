// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen Styles should match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#cca2b7",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 50,
    }
  }
>
  <View>
    <Image
      source={
        {
          "uri": "mockedUrl",
        }
      }
      style={
        {
          "height": 200,
          "marginBottom": 20,
          "width": 200,
        }
      }
    />
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "marginBottom": 10,
        }
      }
    >
      Mocked Track
    </Text>
    <Text
      style={
        {
          "fontSize": 16,
        }
      }
    >
      Artist
    </Text>
  </View>
</View>
`;
